### DISCUSSION TITLE
SQL nvarchar column size

### DISCUSSION THREAD

**Discussion by markgould** [OrchardCMS/OrchardCore#7235] [Q&A] [answered]
Is there a particular reason ContentItemId (and related fields) are 26 chars in sql? We're running into an issue where our recipes work fine in sqlite but not in SQL due to the 26 char limitation.

We are now faced with going through our recipe and changing all the content ids. Is it likely this length could be increased at some point? Thanks!

**Comment by hishamco** [comment]
Why it doesn't work with SQLite? If I'm not wrong the Id is generated by OC itself

**Comment by markgould** [comment]
It does work with sqlite, it does NOT work with MSSQL. SQLite does not have limits on text fields, which is why it works. 

This document lists the 26 char limit: https://github.com/OrchardCMS/OrchardCore/blob/dev/src/docs/reference/modules/SQL/README.md

**Comment by hishamco** [comment]
> Is there a particular reason ContentItemId (and related fields) are 26 chars in sql?

https://github.com/OrchardCMS/OrchardCore/blob/569e256b1aeb86c3321c5233562cea19ee94be9c/src/OrchardCore/OrchardCore.Abstractions/IdGenerator.cs#L14

The Generated Id contains 26 chars

**Comment by markgould** [comment]
Ah ok. We are having issues where people are writing recipes and using their own content ids (alpha numeric) which are > 26 chars.

**Reply by hishamco** [reply]
Aha, if is like that the SQL will complain, IMHO we need to increase the size, because the `DefaultIdGenerator` is using 26 char, but a custom `IIdGenerator` may use more than that

Note for myself while the Id is using in the URL we should take the URL limits in our consideration

/cc @jtkech @sebastienros

**Comment by Skrypt** [comment]
No we should not increase the size. Use the proper UID generator code simply.
`[js: uuid()]` in recipes.
Or use `IContentItemIdGenerator` in server side code.

**Reply by hishamco** [reply]
The `IContentItemIdGenerator` implementation could have different size of what it's there out of the box

**Reply by Skrypt** [reply]
You're right. You should use the OrchardCore/src/OrchardCore/OrchardCore.Abstractions/IdGenerator.cs class.
No reason to change the column size unless what he's saying is that they are using `[js: uuid()]` ans that it would generate id's longer than 26 chars which I'd be surprised of.

If the issue you are facing is to now migrate the data from SQLite to MSSQL truncating the guid might not be much of a big issue since there is really low chances that you get 2 guids that are the same still.

**Reply by Skrypt** [reply]
Or you can use this SQL to regenerate the UUID's properly beforehand. 

```SQL

// SQL Server
LEFT(LOWER(REPLACE(NEWID(), '-', '')), 26)

// SQLite
lower(hex(randomblob(13)))
```

**Comment by jtkech** [comment]
@hishamco 

For info with Kestrel the max url size is 8KB, with IIS/Express + Kestrel the max url **segment** size is 260 by default, but can be [increased through the registry](https://docs.microsoft.com/en-us/troubleshoot/iis/httpsys-registry-windows).

**Comment by deanmarcussen** [comment]
We won't be increasing the size of the column by default.

It will affect db index key length limits on other databases (MySql < 768 vs SQL server < 1024)

If you want to use greater than 26 characters in your implementation you can alter the column length yourself, either from a custom migration or with sql command directly.

Just be careful you don't get warnings about key length limits when you make the change (would recommend doing it through SQL Management Studio, as it will give you a warning, if you go over your key length limits)

Note: There is nothing in the codebase (other than an initial migration) forcing you to stick at 26 characters.

**Comment by hishamco** [comment]
If we agree that we don't have a plan to increase the column size by default due the reason that @deanmarcussen mentioned, at least we need to have a note in our docs to avoid such issue in case someone implement his own `IIdGenerator`

**Comment by markgould** [comment, marked as answer]
I dug into our recipe a little more and figured out the issue. Rather than creating variables like:

"variable1": "[js:uuid()]"

They've all been created like:

"somePageContentId": "somePage"

Some of them have exceeded 26 chars. All we need to do is replace all the variable values to generate a uuid and then we should be in good shape.
